### PR TITLE
Add write method for bits types, mmap tests.

### PR DESCRIPTION
**Pull request by tpapp**
1. Add a method for `write(::IO, ::T)` where `T` is a bits type. Previously, this threw a `MethodError`, left this behavior unchanged for other types.

2. `write(io, [x])` could be used for the same effect before, the added method is compatible with that. This is tested by writing out vectors elementwise, and comparing mmapped values.

3. A new test file for mmap is added. (When this is merged, I plan to add other tests for `mmap`.) 

See related [discussion](https://discourse.julialang.org/t/file-for-mmap-with-an-unknown-number-of-structs/6380/3?u=tamas_papp).

### CODE REVIEW

**Comment by quinnj** [comment]
Mmap tests should probably go in the [mmap test file](https://github.com/JuliaLang/julia/blob/master/stdlib/Mmap/test/runtests.jl)

**Comment by tpapp** [comment]
@quinnj : Thanks, I missed that. This is not a test of `mmap` per se, but the new method for write combined with `mmap`. Would it still be OK to put it in the mmap test file you mentioned above?

**Comment by JeffBezanson** [comment]
This seems ok to me. What do you think @vtjnash ?

File: base/io.jl [hunk 1]
```diff
@@ -334,7 +334,14 @@ isreadonly(s) = isreadable(s) && !iswritable(s)
 
 ## binary I/O ##
 
-write(io::IO, x) = throw(MethodError(write, (io, x)))
+function write(io::IO, x::T) where T
+    if isbits(T)
+        unsafe_write(io, Ref(x), sizeof(T))
```
[the remark this answers is not in the document]
Comment (vtjnash) [reply base/io.jl]: And delete method `write(s::IO, x::Union{Int16,UInt16,Int32,UInt32,Int64,UInt64,Int128,UInt128,Float16,Float32,Float64})`

**Comment by vtjnash** [comment]
Yes, this seems ok to me too.

File: base/io.jl (hunk 1 above)
[the remark this answers is not in the document]
Comment (tpapp) [reply base/io.jl]: Thanks for the suggestions, I made the requested changes, except for defining a `write(::IO, ::Ref{T})` method which [already exists](https://github.com/JuliaLang/julia/blob/master/base/io.jl#L349).

**Comment by tpapp** [comment]
Frankly, I forgot about this PR myself, so I am wondering if there is an interest in merging it.

To recap, this is mostly about whether `write(io, x)` should "just work" when `x` is a bits type, taking the layout in memory as its canonical representation. I don't know enough about the internals to say whether this makes sense in general.

If this considered OK, I can dust off this PR.

**Comment by tpapp** [comment]
Friendly bump: I would still like to get this merged if possible.

If there is anything else I can improve, please tell me.

File: base/io.jl [hunk 2]
```diff
@@ -509,13 +507,17 @@ end
...
 write(s::IO, x::Int8) = write(s, reinterpret(UInt8, x))
-function write(s::IO, x::Union{Int16,UInt16,Int32,UInt32,Int64,UInt64,Int128,UInt128,Float16,Float32,Float64})
-    return write(s, Ref(x))
-end
-
 write(s::IO, x::Bool) = write(s, UInt8(x))
 write(to::IO, p::Ptr) = write(to, convert(UInt, p))
 
+function write(io::IO, x::T) where T
+    if isbits(T)
```
Comment (laborg) [inline base/io.jl]: Shouldn't this be `isbitstype(T)` ?  Alternative: `isbits(x)` but it looks like this would reduce performance.

Comment (tpapp) [reply base/io.jl]: Good catch, I did not follow up on the API update.

**Comment by StefanKarpinski** [comment]
@JeffBezanson, @vtjnash: you've both said this seems ok to you. I think one of you needs to make a final decision and merge this—or not.

**Comment by vtjnash** [comment]
We also could add a check that `!datatype_haspadding(T)`. I'm not a particular fan of making the IO dependent on the system, although I realize we already do that for Arrays. I think the canonical representation would pack the result (remove padding), although that would be less performant and more effort (in the case where it haspadding). None of these options needs a generated function.

File: base/io.jl [hunk 3]
```diff
@@ -509,13 +507,17 @@ end
...
-end
-
 write(s::IO, x::Bool) = write(s, UInt8(x))
 write(to::IO, p::Ptr) = write(to, convert(UInt, p))
 
+function write(io::IO, x::T) where T
+    if isbitstype(T)
+        write(io, Ref(x))
+    else
+        throw(MethodError(write, (io, x)))
```
Comment (KristofferC) [inline base/io.jl]: Imo, manually throwing MethodError is almost always confusing. You get a method error but then you see that the method does exist. Why not throw an error that says that only writing bitstype is allowed or something along those lines?

**Comment by tpapp** [comment]
@vtjnash: for my purposes, the point if this PR is to `write` an *ex ante* unknown number of elements in a way I can `mmap` later. I want padding (if applicable) to be compatible with that.

File: base/io.jl (hunk 3 above)
Comment (tpapp) [reply base/io.jl]: Good idea, will do that in the next commit (I will just wait for the padding discussion to resolve).

**Comment by tlnagy** [comment]
Is there any reason that `write` is so limited? 

The generic write function for `AbstractArray`s tests whether a type is an `isbitstype`, which is all well and good:

https://github.com/JuliaLang/julia/blob/5a6c80828698d58405ca38b27628d426665324b5/base/io.jl#L695-L704

but many valid `isbitstype`s fail with a `MethodError` from line: https://github.com/JuliaLang/julia/blob/5a6c80828698d58405ca38b27628d426665324b5/base/io.jl#L674 

**What's the rational for this?**

This quite a problem for https://github.com/tlnagy/TiffImages.jl/issues/76 because I'm often writing types that hit the above `MethodError`, which forces me to `reinterpret` the array as `UInt8`s first. And that ends up hitting https://github.com/JuliaLang/julia/issues/39781 so performance is suffering for writing.

```julia
julia> using Colors

julia> isbitstype(RGB{Float64})
true

julia> write(io, RGB{Float64})
ERROR: MethodError: no method matching write(::IOBuffer, ::Type{RGB{Float64}})
Closest candidates are:
  write(::IO, ::Any) at io.jl:672
  write(::IO, ::Any, ::Any...) at io.jl:673
  write(::Base.GenericIOBuffer, ::UInt8) at iobuffer.jl:442
  ...
Stacktrace:
 [1] write(io::IOBuffer, x::Type)
   @ Base ./io.jl:672
 [2] top-level scope
   @ REPL[58]:1
```

**Comment by vtjnash** [comment]
I think this is replaced by #42593 and #47116. They don't provide precisely the same result, but give the user some options for dealing with padding when determining the canonical form of the bitstypes.

**Pull request closed without merging**